### PR TITLE
Extend SliceNdLayer to use a layer as the size argument

### PR DESCRIPTION
**Pull request by robin-p-schmitt**
Right now, `start` and `size` need to have the same shape when they are given as layers. This suffices for my use cases but we can also add support for different shapes if you think we need them.

### CODE REVIEW

File: returnn/tf/layers/basic.py [hunk 1]
```diff
@@ -859,39 +859,44 @@ class SliceNdLayer(_ConcatInputLayer):
   def __init__(self, start, size, min_size=None, **kwargs):
     """
     :param LayerBase start: (B,...)
-    :param int|None size: if None, it uses the max possible size, and it becomes a dynamic axis
+    :param int|LayerBase|None size: if None, it uses the max possible size,
+    and it becomes a dynamic axis. If LayerBase, it needs to have the same shape as start
```
Comment (albertz) [inline returnn/tf/layers/basic.py:863]:
```suggestion
      and it becomes a dynamic axis
```
Indent by two spaces.

Remove the layer comment. This should not be the case. Use `get_common_data` on `start`/`size` and then `copy_compatible_to`.

File: returnn/tf/layers/basic.py [hunk 2]
```diff
@@ -859,39 +859,44 @@ class SliceNdLayer(_ConcatInputLayer):
...
-        size = tf.maximum(tf.reduce_max(x.batch_shape[x.time_dim_axis] - start_t), min_size)  # scalar
+        size_t = x.batch_shape[x.time_dim_axis] - start_t
       else:
         # make seq_lens compatible with start_t
         seq_lens = expand_multiple_dims(  # e.g. (B,) or (B,1)
           x=seq_lens,
           axes=[-1] * (len(start_t.shape) - len(seq_lens.shape)))
-        size = tf.maximum(tf.reduce_max(seq_lens - start_t), min_size)  # scalar
+        size_t = seq_lens - start_t
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
```
Comment (albertz) [inline returnn/tf/layers/basic.py:886]: I don't understand your naming. `_t` stands for tensor. So `size_t` is a tensor (not e.g. `Data`). But `size` is what then? This is also a tensor. This doesn't make sense.

File: returnn/tf/layers/basic.py [hunk 3]
```diff
@@ -859,39 +859,44 @@ class SliceNdLayer(_ConcatInputLayer):
...
     :param LayerBase start: (B,...)
-    :param int|None size: if None, it uses the max possible size, and it becomes a dynamic axis
+    :param int|LayerBase|None size: if None, it uses the max possible size,
+    and it becomes a dynamic axis. If LayerBase, it needs to have the same shape as start
     :param int|None min_size: if size is None, but we want to have a min-size
     """
     super(SliceNdLayer, self).__init__(**kwargs)
     from returnn.tf.util.basic import where_bc, expand_multiple_dims
     x = self.input_data.copy_as_batch_major()
     seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None  # (B,) or None
```
Comment (albertz) [inline returnn/tf/layers/basic.py:869]: Note, instead of using `seq_lens` directly here, you could also get the dim tag, then use `dyn_size_ext`, and then functions like `copy_compatible_to`, to simplify the `expand_multiple_dims` below, and also have it more generic.

File: returnn/tf/layers/basic.py [hunk 4]
```diff
@@ -951,7 +958,10 @@ def get_dep_layers(self):
...
     :rtype: list[LayerBase]
     """
-    return super(SliceNdLayer, self).get_dep_layers() + [self.start]
+    dep_layers = super(SliceNdLayer, self).get_dep_layers() + [self.start]
+    if isinstance(self.size, LayerBase):
+      dep_layers += [self.size]
+    return dep_layers
 
   @classmethod
   def get_out_data_from_opts(cls, name, sources=(), start=None, size=None, **kwargs):
```
Comment (albertz) [inline returnn/tf/layers/basic.py:967]: `size` doc should be updated here as `:param int|LayerBase|None size:`.

File: returnn/tf/layers/basic.py (hunk 2 above)
Comment (robin-p-schmitt) [reply returnn/tf/layers/basic.py:886]: Size is either a tensor (in case we use `tf.maximum`) or an integer (if the user provides a static size). Should we maybe rename `size` to `max_size`?

Comment (albertz) [reply returnn/tf/layers/basic.py:886]: Ah sorry, yes. Or also a `LayerBase`. No, leave it like it is.

File: returnn/tf/layers/basic.py [hunk 5]
```diff
@@ -859,39 +859,44 @@ class SliceNdLayer(_ConcatInputLayer):
...
+    and it becomes a dynamic axis. If LayerBase, it needs to have the same shape as start
     :param int|None min_size: if size is None, but we want to have a min-size
     """
     super(SliceNdLayer, self).__init__(**kwargs)
     from returnn.tf.util.basic import where_bc, expand_multiple_dims
     x = self.input_data.copy_as_batch_major()
     seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None  # (B,) or None
     self.start = start
+    self.size = size
     start_data = start.output.copy_as_batch_major()  # e.g. (B,) or (B,T)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:872]: You use `copy_as_batch_major` here and also for `size.output` below.
Do we really need it as batch major? Why? (Please add such a comment in the code.)
Or do we just need it compatible to the input and to `size`? Then we better should use the `get_common_data` and `copy_compatible_to` logic.

This should also be fixed/clarified accordingly in `get_out_data_from_opts`.

**Comment by albertz** [comment]
Btw, also check PyCharm warnings (and corresponding test).
```
  returnn/tf/layers/basic.py:899: WARNING PyUnboundLocalVariableInspection: Local variable 'size_t' might be referenced before assignment
```

File: returnn/tf/layers/basic.py [hunk 6]
```diff
@@ -859,39 +859,44 @@ class SliceNdLayer(_ConcatInputLayer):
...
           x=seq_lens,
           axes=[-1] * (len(start_t.shape) - len(seq_lens.shape)))
-        size = tf.maximum(tf.reduce_max(seq_lens - start_t), min_size)  # scalar
+        size_t = seq_lens - start_t
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
+    elif isinstance(size, LayerBase):
+      assert size.output.batch_ndim == start_data.batch_ndim
+      min_size = 0
+      size_t = size.output.get_placeholder_as_batch_major()
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
```
Comment (albertz) [inline returnn/tf/layers/basic.py:891]: There is some `else:` missing which assigns `size_t`, right? (See also warning)

Comment (albertz) [reply returnn/tf/layers/basic.py:891]: Ok, maybe `size_t` is never used in this case. But anyway assign `None`.

File: returnn/tf/layers/basic.py [hunk 7]
```diff
@@ -859,70 +859,82 @@ class SliceNdLayer(_ConcatInputLayer):
...
         placeholder=dyn_size,
-        dim_tags=start_data.dim_tags,
+        dim_tags=start_data.copy_as_batch_major().dim_tags,  # as batch major because the placeholder is too
         batch=slice_tag.batch,
         beam=slice_tag.batch.beam if slice_tag.batch else self.output.beam,
         control_flow_ctx=slice_tag.control_flow_ctx)
       slice_tag.dyn_size_ext = dyn_size_ext
       slice_tag.set_tag_on_size_tensor(dyn_size)
-    gather_positions_data = start_data.copy_template(name="%s_gather_positions" % self.name)
+    # get as batch major, because the placeholder later will also be batch major
```
Comment (albertz) [inline returnn/tf/layers/basic.py:915]: Why is the placeholder batch major? The placeholder is the same as `start` (`start_t`), or not?

File: returnn/tf/layers/basic.py [hunk 8]
```diff
@@ -859,70 +859,82 @@ class SliceNdLayer(_ConcatInputLayer):
...
+    gather_positions_data = start_data.copy_template(name="%s_gather_positions" % self.name).copy_as_batch_major()
     gather_positions_data = gather_positions_data.copy_add_dim_by_tag(
       slice_tag,
       unbroadcast=True,
       axis=start_data.batch_ndim)
     # [start+0, start+1, ...]
     gather_positions = tf.expand_dims(start_t, -1) + tf.range(0, size)  # e.g. (B, size) or (B, T, size)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:922]: If `start` is e.g. [T,B], then this is [T,B,size], i.e. not batch major. Right?

File: returnn/tf/layers/basic.py [hunk 9]
```diff
@@ -859,70 +859,82 @@ class SliceNdLayer(_ConcatInputLayer):
...
+      size_data = size.output.copy()
+      common_data = Data.get_common_data([start_data, size_data])
+      size_data = size_data.copy_compatible_to(common_data)
+      size_t = size_data.get_placeholder_as_batch_major()
+      start_data = start_data.copy_compatible_to(common_data)
+      start_t = start_data.get_placeholder_as_batch_major()
+      min_size = 0
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
+    else:
+      # in this case, size_t is never used but needs to be set to avoid a PyCharm warning
```
Comment (albertz) [inline returnn/tf/layers/basic.py:895]: Do not add such comment.

File: returnn/tf/layers/basic.py (hunk 8 above)
Comment (robin-p-schmitt) [reply returnn/tf/layers/basic.py:922]: But in the beginning, I set `start_t = start_data.get_placeholder_as_batch_major()`. Should I change that?

Comment (albertz) [reply returnn/tf/layers/basic.py:922]: Why do you use `get_placeholder_as_batch_major` there. If there is no reason for it, yes, remove that. Otherwise comment why it is needed.

File: returnn/tf/layers/basic.py [hunk 10]
```diff
@@ -859,39 +859,48 @@ class SliceNdLayer(_ConcatInputLayer):
...
     from returnn.tf.util.basic import where_bc, expand_multiple_dims
-    x = self.input_data.copy_as_batch_major()
-    seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None  # (B,) or None
+    from returnn.tf.util.data import Data
+    x = self.input_data.copy()
+    seq_lens_data = x.get_time_dim_tag().dyn_size_ext  # (B,) or None
     self.start = start
-    start_data = start.output.copy_as_batch_major()  # e.g. (B,) or (B,T)
+    self.size = size
+    start_data = start.output.copy()  # e.g. (B,) or (B,T)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:873]: We potentially have 3 sources here for the shape of the positions:
`seq_lens_data`, `start` and `size`.
I think we should use `get_common_data` on all three, and then store sth like `common_data = Data.get_common_data([seq_lens_data, start, size.output if isinstance(size, LayerBase) else None])`, and below, use `copy_compatible_to(common_data, ...)`.

File: returnn/tf/layers/basic.py [hunk 11]
```diff
@@ -859,39 +859,48 @@ class SliceNdLayer(_ConcatInputLayer):
...
+      min_size = 0
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
+    else:
+      size_t = None
     # for each start index in start_data, we want to gather a slice
     # therefore, the output's first axes are the same as the ones from start_data
     # and the next axis will therefore be the slice axis
     slice_tag = self.output.dim_tags[start_data.batch_ndim]
     assert slice_tag.description.startswith("sliced-time:")
     if not isinstance(size, int):
```
Comment (albertz) [inline returnn/tf/layers/basic.py:901]: It might be cleaner to check `if size_t is not None:` and then add `assert not isinstance(size, int)`.

File: returnn/tf/layers/basic.py (hunk 10 above)
Comment (robin-p-schmitt) [reply returnn/tf/layers/basic.py:873]: So `get_common_data` does not allow the list items to be `None`. Can I instead do:

```
data_objs = [start_data]
data_objs += [size.output] if isinstance(size, LayerBase) else []
data_objs += [seq_lens_data] if isinstance(seq_lens_data, Data) else []
common_data = Data.get_common_data(data_objs)?
```

Comment (albertz) [reply returnn/tf/layers/basic.py:873]: Ah yes.

**Comment by albertz** [comment]
Note again the code warning:
```
  returnn/tf/layers/basic.py:867: WARNING PyUnresolvedReferencesInspection: Unused import statement <code>expand_multiple_dims</code>
```

**Review by albertz — APPROVED** [review]
Looks good now

**Comment by albertz** [comment]
Just for reference: Related was #661 and #635
